### DISCUSSION TITLE
excluding a field in response works with pydantic base model but not with dataclasses

### DISCUSSION THREAD

**Discussion by mr-mapache** [fastapi/fastapi#12065] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
### SQLAlchemy Repository:### SQLAlchemy Repository:

from typing import Optional
from uuid import UUID, uuid4
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column
from sqlalchemy import ForeignKey

class Schema(DeclarativeBase):
    surrogate_key: Mapped[int] = mapped_column('id', primary_key=True)

class User(Schema):
    __tablename__ = 'users'
    id: Mapped[UUID] = mapped_column('user_id', unique=True, nullable=False)
    username: Mapped[str] = mapped_column('user_username', unique=True, nullable=False)

class Users:
    async def get(self, id: UUID) -> Optional[User]:
       return User(id=UUID("123e4567-e89b-12d3-a456-426614174000"), username='example', surrogate_key=1)

### Does not work##################

from dataclasses import dataclass

##################################

### Works with this:
from pydantic.dataclasses import dataclass

#################################

@dataclass
class User:
    id: UUID
    username: str

######################################

from fastapi import FastAPI, Depends, status, HTTPException

async def users_repository() -> Users:
    return Users()

api = FastAPI()

@api.get("/users/{id}")
async def get_user_by_username(id: UUID, users: Users = Depends(users_repository)) -> User:
    user = await users.get(id)
    if not user:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="User not found")
    return user


from typing import AsyncGenerator
from httpx import AsyncClient, ASGITransport

async def client(users: Users) -> AsyncGenerator[AsyncClient, None]:
    api.dependency_overrides[users_repository] = lambda: users
    test = FastAPI()
    test.mount("/api", api)
    async with AsyncClient(transport=ASGITransport(test), base_url="http://test") as client:
        response = await client.get("/api/users/123e4567-e89b-12d3-a456-426614174000")
        assert response.status_code == status.HTTP_200_OK
        print(response.json())
        assert response.json() == {"id": "123e4567-e89b-12d3-a456-426614174000", "username": "example"}


if __name__ == '__main__':
    import asyncio
    asyncio.run(client(Users()))
```

### Description

I have a domain model written with dataclasses, and a sqlalchemy implementation of a repository written in sqlalchemy. The sqlalchemy repository returns objects with a surrogate key field that is in all schemas, but they are not in domain dataclasses. The response model is set as the dataclass without the surroage key. FastAPI is failing in convert the python dataclass to the pydantic response model excluding the field that is not in the response model. 

Also I do not want to couple my domain layer to pydantic, currently this works:

### Does not work
```python
from dataclasses import dataclass

@dataclass
class User:
    id: UUID
    username: str
```

### Works with this:
```python
from pydantic.dataclasses import dataclass

@dataclass
class User:
    id: UUID
    username: str
```

I think fastapi should not have a different behavior with pydantic dataclasses and python dataclasses. 

### Operating System

Linux

### Operating System Details

wsl 24.04LTS

### FastAPI Version

0.112.1

### Pydantic Version

 2.8.2

### Python Version

Python 3.12.3

### Additional Context

Obviously this is just working a minimal example, i have decoupled the above in different files and has more code.

**Comment by YuriiMotov** [comment, marked as answer]
Thanks for reporting this!

It's said in the docs that although `dataclass`'es will be automatically converted to Pydantic models, in some cases you still need to use `pydantic.dataclasses.dataclass` explicitly (https://fastapi.tiangolo.com/advanced/dataclasses/#dataclasses-in-nested-data-structures).


If you import those classes and can't replace `dataclasses` with `pydantic.dataclasses`, you can use the following trick:

```python
class UserSchemaPD(pydantic.BaseModel, UserSchema):
    pass

@api.get("/users/{id}", response_model=UserSchemaPD)
async def get_user_by_username(id: UUID, users: Users = Depends(users_repository)):
    user = await users.get(id)
    if not user:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="User not found")
    return user

```
